nalgebra-sparse/src/pattern.rs

```diff
@@ -184,6 +186,131 @@ impl SparsityPattern {
         })
     }
 
+    /// Try to construct a sparsity pattern from the given dimensions, major offsets
+    /// and minor indices.
+    ///
+    /// Returns an error if the data does not conform to the requirements.
+    pub fn from_offset_and_indices_unchecked(
+        major_dim: usize,
+        minor_dim: usize,
+        major_offsets: Vec<usize>,
+        minor_indices: Vec<usize>,
+    ) -> Result<Self, SparsityPatternFormatError> {
+        use SparsityPatternFormatError::*;
+
+        if major_offsets.len() != major_dim + 1 {
+            return Err(InvalidOffsetArrayLength);
+        }
+
+        // Check that the first and last offsets conform to the specification
+        {
+            let first_offset_ok = *major_offsets.first().unwrap() == 0;
+            let last_offset_ok = *major_offsets.last().unwrap() == minor_indices.len();
+            if !first_offset_ok || !last_offset_ok {
+                return Err(InvalidOffsetFirstLast);
+            }
+        }
+
+        Ok(Self {
+            major_offsets,
+            minor_indices,
+            minor_dim,
+        })
+    }
+
+    /// Validates cs data, sorts minor indices and values
+    pub fn validate_and_optionally_sort_cs_data<T>(
+        major_dim: usize,
+        minor_dim: usize,
+        major_offsets: &[usize],
+        minor_indices: &mut [usize],
+        values: Option<&mut [T]>,
+        sort: bool,
+    ) -> Result<(), SparsityPatternFormatError>
+    where
+        T: Scalar,
+    {
+        use SparsityPatternFormatError::*;
+
+        if major_offsets.len() != major_dim + 1 {
+            return Err(InvalidOffsetArrayLength);
+        }
+
+        // Check that the first and last offsets conform to the specification
+        {
+            let first_offset_ok = *major_offsets.first().unwrap() == 0;
+            let last_offset_ok = *major_offsets.last().unwrap() == minor_indices.len();
+            if !first_offset_ok || !last_offset_ok {
+                return Err(InvalidOffsetFirstLast);
+            }
+        }
+
+        let mut minor_index_permutation: Vec<usize> = (0..minor_indices.len()).collect();
```

Comment:
*Always* allocating a vector here is very unfortunate. We need to make sure we don't have any heap allocations on the "happy path" (i.e. when the data is already in a valid format). One way to ensure this is to instead think of the permutation as a buffer, and then only permute *locally*. Roughly like this:

- Start with empty vector `Vec::new()`
- *If* we should sort and a lane is unsorted, then clear and populate the permutation buffer with `(0 .. lane_size)`
- Apply the permutation to the minor indices, and possibly the values if they are provided

This way the vector stays empty if there is no need to sort (and thus never needs to heap allocate). Makes sense?

Reply:
I think it makes sense and try to apply this idea.

Reply:
I reduced allocations of permutations in `validate_and_optionally_sort_cs_data`. My implementation might deviate a bit from your idea description but I hope it does roughly the same thing. :)

Further, I think the expectation on `validate_and_optionally_sort_cs_data` could be more clearer.
At the moment, if sorting is expected then values should be provided as well and an error is thrown otherwise.
Or should this behavior be changed?
Maybe we could even get rid of the boolean `sort` argument and let the `validate_and_optionally_sort_cs_data` sort data always when given indices are unsorted.
If we use `validate_and_optionally_sort_cs_data` without expecting it to sort data then its functionality seems to be redundant. One could use `try_from_csr_data`/`try_from_csc_data` instead.